File: client/src/lib/provider-errors.ts
/** Map provider / API error text into localized, user-facing copy. */

type Translate = (key: string) => string

function normalizeErrorText(raw: string): string {
  return raw
    .replace(/^(generate_master|analyze uploaded master|transition generate|panel\[\d+\] process|transition panel\[\d+\]) failed:\s*/i, "")
    .trim()
}

export function formatApiError(error: unknown, t: Translate): string {
  const raw = error instanceof Error ? error.message : String(error || "")
  const text = normalizeErrorText(raw)
  const lower = text.toLowerCase()

  if (
    lower.includes("does not support reference editing") ||
    lower.includes("does not support editing") ||
    lower.includes("text-to-image only")
  ) {
    return t("errors.editingUnsupported")
  }

  if (
    lower.includes("authentication") ||
    lower.includes("api key rejected") ||
    lower.includes("api key does not match") ||
    lower.includes("key must match") ||
    lower.includes("key doesn’t match") ||
    lower.includes("key doesn't match") ||
    lower.includes("authentication_error") ||
    lower.includes("unauthorized") ||
    lower.includes("ai_gateway_api_key") ||
    lower.includes("invalid api key") ||
    lower.includes("invalid_api_key")
  ) {
    return t("errors.keyModelMismatch")
  }

  if (lower.includes("rate limit") || lower.includes("too many requests")) {
    return t("errors.rateLimited")
  }

  if (
    lower.includes("quota") ||
    lower.includes("balance exhausted") ||
    lower.includes("insufficient") ||
    text.includes("余额") ||
    text.includes("额度") ||
    text.includes("欠费")
  ) {
    return t("errors.quotaExhausted")
  }

  if (
    lower.includes("model") &&
    (lower.includes("not found") ||
      lower.includes("does not exist") ||
      lower.includes("unknown model") ||
      lower.includes("invalid model") ||
      lower.includes("not support") ||
      lower.includes("invalid for this endpoint"))
  ) {
    return t("errors.invalidModel")
  }

  if (
    lower.includes("no image generated") ||
    lower.includes("did not return any images") ||
    lower.includes("missing image")
  ) {
    return t("errors.noImage")
  }

  if (
    lower.includes("ai gateway request failed") ||
    (lower.includes("provider ") && lower.includes(" failed")) ||
    lower.includes("image request failed") ||
    lower.includes("openai image request failed")
  ) {
    return t("errors.providerFailed")
  }

  // Prefer first line; avoid dumping provider JSON into the UI
  const first = text.split("\n").map((s) => s.trim()).find(Boolean) || text.trim()
  if (!first) return t("errors.providerFailed")
  if (first.startsWith("{") || first.startsWith("[")) return t("errors.providerFailed")
  return first
}

File: server/app/services/provider_errors.py
"""Map provider HTTP error bodies to short, user-facing messages."""

from __future__ import annotations


def friendly_provider_error(
    body: str | None,
    *,
    status_code: int = 500,
    what: str = "request",
) -> str:
    """Return a clear message for the `error` field (keep raw body in `details`)."""
    text = (body or "").strip()
    lower = text.lower()

    authish = (
        status_code in (401, 403)
        or "authentication" in lower
        or "unauthorized" in lower
        or "invalid api key" in lower
        or "invalid_api_key" in lower
        or ("api key" in lower and ("invalid" in lower or "incorrect" in lower or "failed" in lower))
        or "ai_gateway_api_key" in lower
    )
    if authish:
        return (
            "API key rejected. Key must match the selected model/endpoint: "
            "Zhipu key → CogView models; Gemini/OpenAI → Vercel AI Gateway key "
            "(or set a matching proxy Base URL in Workspace)."
        )

    if status_code == 429 or "rate limit" in lower or "too many requests" in lower:
        return "Provider rate limit reached. Please wait a moment and try again."

    if (
        "model" in lower
        and (
            "not found" in lower
            or "does not exist" in lower
            or "unknown model" in lower
            or "invalid model" in lower
            or "not support" in lower
        )
    ):
        return (
            "Model is invalid for this endpoint. "
            "Open Workspace and pick a model that matches your API key."
        )

    if status_code >= 500:
        return f"Provider {what} failed (server error). Please retry shortly."

    return f"Provider {what} failed. Check Workspace model and API key settings."
